src/SymfonyCache/KernelClient.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCache package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCache\SymfonyCache;
+
+use Http\Client\HttpAsyncClient;
+use Http\Promise\FulfilledPromise;
+use Psr\Http\Message\RequestInterface;
+use Symfony\Bridge\PsrHttpMessage\Factory\DiactorosFactory;
+use Symfony\Bridge\PsrHttpMessage\Factory\HttpFoundationFactory;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Zend\Diactoros\ServerRequest;
+
+/**
+ * An implementation of HttpAsyncClient that allows direct routing through the
+ * Symfony HttpCache kernel without executing a real HTTP request.
+ * It uses the HttpFoundationFactory and the Zend DiactorosFactory to convert
+ * between PSR-7 requests and responses. Both are optional dependencies of this
+ * package thus existence of the respective classes is checked for in the
+ * constructor of this client. It is only needed if you have a single node
+ * setup of Symfony and serves as kind of a shortcut for easier configuration.
+ * If you use Varnish or have a multiple node Symfony setup, this class is entirely
+ * useless to you and you can happily ignore it.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class KernelClient implements HttpAsyncClient
+{
+    /**
+     * @var HttpCacheAwareKernelInterface
+     */
+    private $kernel;
+
+    /**
+     * @var HttpFoundationFactory
+     */
+    private $httpFoundationFactory;
+
+    /**
+     * @var DiactorosFactory
+     */
+    private $psr7Factory;
+
+    /**
+     * KernelClient constructor.
+     *
+     * @param HttpCacheAwareKernelInterface $kernel
+     */
+    public function __construct(HttpCacheAwareKernelInterface $kernel)
+    {
+        $this->kernel = $kernel;
+
+        if (!class_exists(HttpFoundationFactory::class)) {
+            throw new \RuntimeException('Install symfony/psr-http-message-bridge to use this client.');
+        }
+
+        if (!class_exists(DiactorosFactory::class)) {
+            throw new \RuntimeException('Install zendframework/zend-diactoros to use this client.');
+        }
+
+        $this->httpFoundationFactory = new HttpFoundationFactory();
+        $this->psr7Factory = new DiactorosFactory();
+    }
+
+    /**
+     * Converts a PSR-7 request to a Symfony HttpFoundation request, runs
+     * it through the HttpCache kernel and converts the response back to a PSR-7
+     * response.
+     *
+     * {@inheritdoc}
+     */
+    public function sendAsyncRequest(RequestInterface $request)
+    {
+        $symfonyRequest = Request::createFromGlobals();
```

Comment:
should we not do `new Request` here? otherwise we might sneak accidental things from whatever request was going on that is triggering the invalidation.

Reply:
Yeah wasn't sure about that. I've used `new Request()` first but then it's just probably missing a lot of other info (although that is pretty surely entirely irrelevant info for the cache purging process).

Reply:
we don't do that in the other invalidation requests, so i think it should be safe. if we need specific things, we should get the original request from the RequestStack service and copy the things. (Host header should probably be present, for example)

Reply:
There is no request stack at this point. Well there is but no current request. We are talking about the `kernel.terminate` event there. I used to have this in the first iteration but then noticed that and thus my fallback was the globals :)